### DISCUSSION TITLE
Redmine SLA plugin

### DISCUSSION THREAD

**Discussion by maggne** [likehopper/redmine_sla#13] [Q&A]
Hello, is there some way how to connect 3 different Service levels with 3 different calendars in one tracker?

For example, I have created 3 levels.
1st one - working hours 24/7
2nd - working hours 8am to 20pm
3rd -  working hours 9am - 17pm

I have created 3 calendars with this working hours, combinated with those levels.
After setting it all, the priorities in terms, statuses, everything is set correctly I guess, the issues I created, showed only one level and calendar( always the one which was created first). It should be changed always when I change the priority, but it never works.

Is there any possibility to do that?
thanks for responding

![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/7a8bd9ad-dba1-4577-8834-ce6e16cc3f20)

![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/116136de-7d22-4a91-9f8c-99b8ca9b2ca6)
![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/be8f2e29-0185-4e1b-b3f8-c6567b7781b8)
![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/11e849fb-1a6c-4947-a15e-7e6b6a06c7fc)

**Comment by likehopper** [comment]
Thank you for your interest in the plugin.
This should work unless the schedulers overlap?
In the project settings, what SLAs are assigned to which trackers?
Can you try exporting SLA data for me with these Python scripts for example ?
[SLA-import-export.zip](https://github.com/likehopper/redmine_sla/files/14085169/SLA-import-export.zip)

**Reply by maggne** [reply]
All three SLAs are assigned to only one tracker (Incident) and we are dividing the issues between those 3 SLAs (A, B, C).
So now I think it is not working because those schedulers are overlapping... unfortunately.
I can try to export it tomorrow.
Thank you for replying.

**Reply by maggne** [reply]
Any update?

**Reply by likehopper** [reply]
Can I get your export?

**Reply by maggne** [reply]
We found other way, how to do it. Unfortunately.. for example the Reaction time doesnt show up after changing statuses.
It keeps starting after one hour! At SLA plugin setting the time-zone is correct. Don't know why. Our RM server time is same as our time-tone. 
Here is the export. Would be glad for your tips.
[SLA-plug-export.txt](https://github.com/likehopper/redmine_sla/files/14258753/SLA-plug-export.txt)

**Reply by likehopper** [reply]
Thanks, I got the export, I'll import it and take a look.

**Reply by maggne** [reply]
Okay, thank you.

**Reply by likehopper** [reply]
I just imported the file. I will be able to start testing.

**Reply by maggne** [reply]
Okay. Let me know. Thanks

**Comment by likehopper** [comment]
I understood why: because there is only one event in the history of the issue. Just change the status and the SLAs will appear. This comes from an recent optimization...

**Reply by likehopper** [reply]
Could you try updating the database definition with the view below?

```
CREATE OR REPLACE VIEW sla_view_journal_statuses AS

    SELECT DISTINCT issues.id AS issue_id,
        sla_get_date(first_value(issues.created_on) OVER window_journals) AS issue_created_on,
        sla_get_date(first_value(issues.closed_on) OVER window_journals) AS issue_closed_on,
        sla_get_date(first_value(issues.created_on) OVER window_journals) AS journals_created_on,
        COALESCE(first_value(journal_details.old_value::integer) OVER window_journals, issues.status_id) AS journal_detail_old_value,
        COALESCE(first_value(journal_details.old_value::integer) OVER window_journals, issues.status_id) AS journal_detail_value
    FROM issues
    LEFT JOIN journals ON ( issues.id = journals.journalized_id )
    LEFT JOIN journal_details ON ( journals.id = journal_details.journal_id AND journal_details.property LIKE 'attr' AND journal_details.prop_key LIKE 'status_id' )
    WINDOW window_journals AS (PARTITION BY issues.id ORDER BY journal_details.id ASC NULLS LAST )

UNION

    SELECT issues.id AS issue_id,
        sla_get_date(issues.created_on) AS issue_created_on,
        sla_get_date(issues.closed_on) AS issue_closed_on,
        sla_get_date(journals.created_on) AS journals_created_on,
        journal_details.old_value::integer AS journal_detail_old_value,
        journal_details.value::integer AS journal_detail_value
    FROM issues
    INNER JOIN journals ON ( issues.id = journals.journalized_id )
    INNER JOIN journal_details ON ( journals.id = journal_details.journal_id AND journal_details.property LIKE 'attr' AND journal_details.prop_key LIKE 'status_id' )

```

**Reply by maggne** [reply]
Hello, thanks for updating. I'll let my colleague know and we will try to update it. 
I'll let you know the result

**Reply by maggne** [reply]
Hello, 
after updating the problem is still here.
Looks like there is a problem with time-zone.

After setting up our time-zone (Prague +1) the SLA reaction time starts still after one hour and the resolving time shows one hour difference, always. 
We also tested UTC time-zone also, but the problem is still the same.

Can you check the time-zone settings? Or the time counting in SLA?

Thank you

**Comment by likehopper** [comment]
Thank you for your patience and your contribution to the improvement of this plugin.

Time zones are indeed complicated to manage because they are set at the server, database and Redmine level. Sorry, I lack perspective on the subject and it is possible that this needs to be refined! Would it be possible to carry out the tests on the same recommended configuration entirely in 'UTC'?

First, change the zone of the linux' server :

	This is my server config with local time :  
```
		# timedatectl | grep 'Time zone'
		        Time zone: Europe/Paris (CET, +0100)
		# cat /etc/timezone
			Europe/Paris
```
	So I change for UTC with this command : 
```
		# timedatectl set-timezone UTC
```
	It's like this command :
```
		# ln -sf /usr/share/zoneinfo/UTC /etc/localtime
```
	And then then the local time is correct :   
```
		# timedatectl | grep 'Time zone'
		        Time zone: UTC (UTC, +0000)
```
	You can ever change timz zone with this command :
```
		# dpkg-reconfigure tzdata
```
		
	And then the time zone is correct too :
```
		# cat /etc/timezone 
		UTC
```

Second, edit 'postgresql.conf' to set 'log_timezone' and 'timezone' to 'Etc/UTC' and then restart postgres' server :
```
	# grep 'timezone' /etc/postgresql/13/main/postgresql.conf 
		log_timezone = 'Etc/UTC'
		timezone = 'Etc/UTC'
	$ $ psql redmine -c "SHOW TIME ZONE;"
 		TimeZone 
		----------
 		Etc/UTC
		(1 ligne)
```

Third, make sure you have also selected 'etc/UTC' in Redmine settings. 

Are new tests conclusive? Otherwise, would you need to give me the configuration of the time zones of your server, your database and Redmine?


_NB : If necessary, it is possible to clear the cache manually with this command:
	# psql redmine -c "TRUNCATE sla_caches CASCADE; "_

**Reply by maggne** [reply]
Hello,
thank you for quick response!

We will try to change all the settings which you mentioned and I will let you know the result.
Thank you once again.

**Reply by maggne** [reply]
Hello,
after last the last change (https://github.com/likehopper/redmine_sla/discussions/13#discussioncomment-8683538)
The SLA plugin is not working at all. We have tried to clear the cache twice and also set up all the SLA values again.

Everything is saved without any problems or errors. Also we tried to check Redmine logs if there is some error message - NONE!.

It is acting like it is working without any problem, but when I create an issue, the SLA do not appear. (Correct status and priority - checked many times.) 
Could it be caused by the last update?
![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/e7dc0bf3-b46d-4730-8505-45528b81bd17)

**Comment by likehopper** [comment]
Changing the timezone configuration shouldn't invalidate the plugin. For my part, I was in “Europe/Paris” and I also switched to “Etc/UTC” without any problem.

It's indeed necessary to check the status and priority. But, it's also necessary to ensure that the SLA is well defined for this tracker in the project configuration:
![Capture d’écran du 2024-03-12 22-59-44](https://github.com/likehopper/redmine_sla/assets/12208142/23ae4fb9-ff35-44ba-8ca4-ba5944ac42e5)

The display of the SLA also depends on the user's role in the project. So, heck the role rights in “Roles and Permissions” in “Administration”. For example, for a manager, the rights can be:
![Capture d’écran du 2024-03-12 23-06-54](https://github.com/likehopper/redmine_sla/assets/12208142/3b6a31a1-28d8-41a2-894e-75c1f8e76504)

Otherwise, to allow me to reproduce the issue again, could you send me the SLA export?
And possibly, could I have an export of the database to see what's happening in your case?


NB : The attached script can be used to anonymize the database (excluding plugins).
[redmine_make_anonymous.txt](https://github.com/likehopper/redmine_sla/files/14579334/redmine_make_anonymous.txt)

**Reply by maggne** [reply]
Yeah, I see. I followed your SLA/doc examples, so I think I've set it all correctly.
There you can see Statuses:
![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/214bf7e4-d00f-4543-85ea-84257f4d27d9)

And priorities, which we are using in our ticket system:
![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/91ace214-b5f7-402f-9361-b9debaaa392a)

Roles and permissions are also set correctly. As I said, I followed your examples many times, it was everything OK, until we changed the time-zone on Database/server level.

I'll send you those export ASAP.
Thank you for your willingness to help.

**Reply by maggne** [reply]
Hello, 
so here im sending you exports which you wanted.

**SLA plugin export:**
[SLAplugExp.txt](https://github.com/likehopper/redmine_sla/files/14649192/SLAplugExp.txt)

The Redmine DB export will be sent on mail, because the file is bigger than it is allowed here.

**Reply by maggne** [reply]
Could you leave an email here? So I can send you the next export? Thank you.

**Reply by likehopper** [reply]
Could you try to send it to <friwaduxero-9082@yopmail.com>, please?

**Reply by likehopper** [reply]
And could you try updating the database definition with the view below in the meantime ?
```
CREATE OR REPLACE VIEW sla_view_roll_statuses AS
(
  -- The first and subsequent status changes
	SELECT
	  issue_id AS issue_id,
	  journal_detail_old_value AS from_status_id,
	  LAG(journals_created_on,1,issue_created_on) OVER window_status AS from_status_date,
	  journal_detail_value AS to_status_id,
	  journals_created_on AS to_status_date	FROM sla_view_journal_statuses
	WINDOW window_status AS ( PARTITION BY issue_id ORDER BY journals_created_on ASC )
) UNION (
  -- To always have the last status change with issue closed or now
  SELECT
  	issue_id AS issue_id,
  	FIRST_VALUE(journal_detail_value) OVER window_status AS from_status_id,
  	FIRST_VALUE(journals_created_on) OVER window_status AS from_status_date,
  	FIRST_VALUE(journal_detail_value) OVER window_status AS to_status_id,
  	COALESCE(issue_closed_on,sla_get_date(NOW()::TIMESTAMP WITHOUT TIME ZONE)) AS to_status_date
  FROM sla_view_journal_statuses
  WINDOW window_status AS ( PARTITION BY issue_id ORDER BY journals_created_on DESC )
)
```

**Reply by likehopper** [reply]
For the moment, with the export sent, I am unable to reproduce the issue. I hope the database will allow me to do this.

**Reply by maggne** [reply]
Somehow, I cant send it on the mail you mentioned. Not working from both my mail iam using. Can you try to send me other one?

Do you want us to update the DB with the view or you want to wwait until i will send you the DB export?

**Reply by likehopper** [reply]
ok, you can try with this mail <xxx.xxx@xxx.xxx> to send me the anonymized dump as is ( so without the new view ).
At the same time, you can update the view of your environment on your side

**Reply by maggne** [reply]
Sent it. Thank you
We will update it during the day. I will let you know.

**Reply by likehopper** [reply]
Ok, I was able to import the database.
Just, the 'type' field of the 'custom_fields' table should not be transformed.

**Reply by likehopper** [reply]
I think I reproduced the error with issue #66142 in project #224. I can now begin my investigations.

**Reply by likehopper** [reply]
Could you try checking the “match” field?
![image](https://github.com/likehopper/redmine_sla/assets/12208142/a85916f8-3b2a-4ee7-8f6e-7a56323ab1c6)

**Reply by maggne** [reply]
Okay. Now it seems that it is working perfectly!! 
I have to test it with more statuses and differents situations. 

**Btw. I checked those "match" fields, but doesnt it mean that the SLA will be running after working hours? (8-20)**

**Reply by maggne** [reply]
Is there any option that the SLA on tickets could be called via API HTTP requests?

Also, do you think you could add one more field like "Deadline", which would show the exact date of resolution?

**Reply by likehopper** [reply]
I try to explain the "match" field here : https://github.com/likehopper/redmine_sla/blob/main/doc/EXAMPLE-03.md

The idea is that the time slots where the incident can be reported then the "match" field is checked.
If the hours are defined but the "match" field is not checked, then the SLA continues to run; but if a new incident occurs in a time slot where the "match" field is not checked, then it will only start at the next time slot where the "match" field is checked.

Am I clear enough in my explanations?

**Reply by likehopper** [reply]
Today the API HTTP requests only allows you to configure and view SLAs.
I'm working on a feature to allow consultation of SLAs by API HTTP requests... I hope to release this at the beginning of April.
I note the field "Deadline" but the difficulty is that it depends on the evolution of the status of the issue.

**Reply by maggne** [reply]
Hello,
the SLA plugin is working fine at the moment. 
But still we need some modifications here. 

Is possible somehow to set that the system field "Priority" would be showing only default values? Not the values we are using via SLA plugin ( for example - our priorities are divided in to 3 levels (A,B,C) and each level is divided in to 4 customer priorities (P1,.. P4). 
So if the Project would not have the SLA in, it would be showing only default values. 

Screen examples:
Project with SLA
![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/4c5a6fe9-28e2-4b62-ace2-a32bd8b952b9)

would be showing all priority options
![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/b0820ea6-945b-40d2-9de6-0a8677c06fa6)

and project without SLA, these default values:
![obrazek](https://github.com/likehopper/redmine_sla/assets/157802929/0a879033-a25d-476d-a602-4720655a692a)

Not sure, if this is quesion on the SLA plugin, rather possibly Redmine settings?

**If, unfortunately, this wasn't possible.. could the Priority field be replaced by any custom field? That would definitely solve our problem described above.** 

Thanks for patience with us.

**Reply by likehopper** [reply]
I created a new discussion (https://github.com/likehopper/redmine_sla/discussions/14) on this subject, because implementing other solutions can quickly become complex.

**Reply by likehopper** [reply]
Why not stay on three priorities (A, B & C) and sub-projects per customer (P1, P2, P3 and P4) distributed across four sub-projects (possibly each with its own SLA) ?

**Comment by likehopper** [comment]
Have you been able to test the version I sent you? I just published a brand new stable version.

**Comment by likehopper** [comment]
Have you tested the latest available version ? Does it meet your needs ?
